File: examples/ccxt.pro/benchmark/ts/src/benchmark.ts
import * as ccxt from 'ccxt';

import {
  wsConnecting,
  wsMsgsReceived,
  wsProcessing,
  wsDisconnects,
  wsSessionDuration,
} from './metrics';

const watchMethodDefaults: Record<string, any[]> = {
  watchOHLCV: ['BTC/USDT', '1m'],
  watchOrderBook: ['BTC/USDT'],
  watchTicker: ['BTC/USDT'],
  // watchTickers: [],
  watchTrades: ['BTC/USDT'],
  //watchBalance: [],
};

export async function benchmarkWatchMethod(exchange: any, methodName: string, args: any[]) {
  console.log(exchange.id, methodName, args);
  const labels = { exchange: exchange.id, method: methodName, language: 'TS'};

  async function executeMethod(sessionStart: number) {
    const start = Date.now();
    try {
      const data = await exchange[methodName](...args);
      const duration = (Date.now() - start) / 1000;
      wsMsgsReceived.labels(labels.exchange, labels.method, labels.language).inc();
      wsProcessing.labels(labels.exchange, labels.method, labels.language).observe(duration);
      return data;
    } catch (error) {
      wsDisconnects.labels(labels.exchange, labels.method, labels.language).inc();
      wsSessionDuration.labels(labels.exchange, labels.method, labels.language)
        .observe((Date.now() - sessionStart) / 1000);
      console.error(`[${labels.exchange}] ${labels.method} error:`, error);
      throw error;
    }
  }

  let sessionStart = Date.now();
  // Continuous monitoring
  while (true) {
    try {
      await executeMethod(sessionStart);
    } catch (error) {
      console.log('error', error);
      sessionStart = Date.now();
    }
  }
}

export function startBenchmarks() {
  //const exchanges = ccxt.exchanges;
  const exchanges = ["binance", "kucoin", "poloniex", "kraken"];
  const CCXT = ccxt as any; // Hack!
  Object.values(exchanges).forEach((exchangeName) => {
    console.log (exchangeName);
    try {
        // @ts-ignore
        const exchange = new CCXT.pro[exchangeName]({}) as ccxt.Exchange
        Object.keys(exchange.has)
        .filter((key) => key.startsWith('watch') && exchange.has[key])
        .forEach((method) => {
            if (watchMethodDefaults[method]) {
              benchmarkWatchMethod(exchange, method, watchMethodDefaults[method]);
            } else {
              console.warn(`No defaults for ${exchange.id} ${method}`);
            }
        });
    } catch (e) {
        console.log (e);
    }
   });
}

File: examples/ccxt.pro/benchmark/python/benchmark.py
import asyncio
import ccxt.pro as ccxtpro
import time
from metrics import *

watch_method_defaults = {
    'watchOHLCV': ['BTC/USDT', '1m'],
    'watchOrderBook': ['BTC/USDT'],
    'watchTicker': ['BTC/USDT'],
    'watchTrades': ['BTC/USDT'],
}

async def benchmark_watch_method(exchange, method_name, args):
    print(exchange.id, method_name, args)
    labels = {'exchange': exchange.id, 'method': method_name, 'language': 'Python'}

    async def execute_method(session_start):
        start = time.time()
        try:
            data = await getattr(exchange, method_name)(*args)
            duration = time.time() - start
            ws_msgs_received.labels(**labels).inc()
            ws_processing.labels(**labels).observe(duration)
            return data
        except Exception as error:
            ws_disconnects.labels(**labels).inc()
            ws_session_duration.labels(**labels).observe(time.time() - session_start)
            print(f"[{labels['exchange']}] {labels['method']} error:", error)
            raise error

    while True:  # Outer loop for session restart
        session_start = time.time()
        try:
            while True:  # Inner loop for continuous monitoring
                await execute_method(session_start)
        except Exception as error:
            print('error, restarting session', error)
            await asyncio.sleep(1)  # Wait 1s before restart

async def start_benchmarks():
    exchanges = list(ccxtpro.exchanges)[:5]  # Start with 5 exchanges
    exchanges = ["binance", "kucoin", "poloniex", "kraken"]
    for exchange_name in exchanges:
        try:
            exchange = getattr(ccxtpro, exchange_name)()
            methods = [
                method for method in dir(exchange)
                if method.startswith('watch') and callable(getattr(exchange, method))
            ]
            
            for method in methods:
                if method in watch_method_defaults:
                    await benchmark_watch_method(
                        exchange,
                        method,
                        watch_method_defaults[method]
                    )
                else:
                    print(f"No defaults for {exchange.id} {method}")
        except Exception as e:
            print(e)
